library/Garden/Web/ResourceRoute.php

```diff
@@ -0,0 +1,351 @@
+<?php
+/**
+ * @author Todd Burry <todd@vanillaforums.com>
+ * @copyright 2009-2017 Vanilla Forums Inc.
+ * @license GPLv2
+ */
+
+namespace Garden\Web;
+
+use Garden\ClassLocator;
+use Interop\Container\ContainerInterface;
+
+/**
+ * Maps requests to controllers using RESTful URLs.
+ *
+ * Here are some of the features of this route.
+ *
+ * - You can attach the route to base path. (ex. /api/controller maps to ApiController).
+ * - You can customize the naming scheme of the controller with a controller pattern.
+ * - Add parameter constraints to help clean data and disambiguate between different endpoints.
+ * - Controllers can be created through an optional dependency injection container.
+ * - Class and method lookup can be customized.
+ * - Supports different controller methods for different HTTP methods.
+ */
+class ResourceRoute extends Route {
+    private static $specialMethods = ['index', 'get', 'post', 'patch', 'put', 'options', 'delete'];
+
+    /**
+     * @var string
+     */
+    private $basePath;
+
+    /**
+     * @var string
+     */
+    private $controllerPattern;
+
+    /**
+     * @var ContainerInterface
+     */
+    private $container;
+
+    /**
+     * @var ClassLocator
+     */
+    private $classLocator;
+
+    /**
+     * Initialize a new {@link ResourceRoute}.
+     *
+     * @param string $basePath The base path to route to.
+     * @param string $controllerPattern A controller pattern that defines how controller classes are named.
+     * @param ContainerInterface|null $container An optional container used to create controller instances.
+     * @param ClassLocator|null $classLocator A class locator used to lookup classes and methods on the controllers.
+     */
+    public function __construct(
+        $basePath = '/',
+        $controllerPattern = '%sController',
+        ContainerInterface $container = null,
+        ClassLocator $classLocator = null
+    ) {
+        parent::__construct();
+        $this->setBasePath($basePath);
+        $this->setControllerPattern($controllerPattern);
+        $this->container = $container;
+        $this->classLocator = $classLocator ?: new ClassLocator();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function match(RequestInterface $request) {
+        $path = $request->getPath();
+
+        // First check and strip the base path.
+        if (stripos($path, $this->basePath) === 0) {
+            $pathPart = substr($path, strlen($this->basePath));
+        } else {
+            return null;
+        }
+
+        $pathArgs = explode('/', $pathPart);
+
+        // First look for the controller.
+        $controllerSlug = $this->filterName(array_shift($pathArgs));
+        if (null === ($controllerClass = $this->classLocator->findClass(sprintf($this->controllerPattern, $controllerSlug)))) {
```

Comment:
Speaking of readability/clarity, I find that `$var === null` is more readable than `null === $var`.